Build scripting

TsubasaK111 opened the pull request:
@takayukioda review me!
- I built a script to automate static page build chores.
- This also implements tag manager on each page.

TsubasaK111 committed:
chore: rebuild static pages

TsubasaK111 committed:
feat: add build script for custom building, WIP

TsubasaK111 committed:
feat: add HTMLparser and search html for modification points

TsubasaK111 committed:
fix: correct usage of Set class instances

TsubasaK111 committed:
feat: inject tagManagerScript successfully

TsubasaK111 committed:
feat: achieve build script functionality :D

TsubasaK111 committed:
chore: remove static files for merge to master

TsubasaK111 committed:
refactor: make build script more readable

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+        self.results          = set([])
+        self.targetIsFullHTML = False
+        self.targetFullPath   = htmlFileFullPath
+        self.pruneThisLang    = set([])
```
I don't know about python culture, but include word "this" feels weird to me. Even if you use "this", because it's a list and could be multiple, `pruneTheseLangs` would be better.
I would use `pruneLangs` or `pruneLangList` instead.

TsubasaK111 replied:
Thanks for the tips, I'll go with `pruneLangs`!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+        if data == "DOCTYPE html":
+            self.targetIsFullHTML = True
+
+            if not ('/en/' in self.targetFullPath or '/ja/' in self.targetFullPath):
```
Is this part needs to be nested?
I think prune language should be independent with doc type.

TsubasaK111 replied:
This snippet is to mark top-level pages to remove TOC items for both languages, and is necessary.
(eg: /index.html, /search.html)

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+                'ja' in self.pruneThisLang and
+                tag == 'a' and
+                attr[0] == 'href' and
+                ('/ja/' in attr[1] or attr[1].startswith('ja/'))
```
Too long if statement would make us confuse, so simplify it by group this as function would be better.

TsubasaK111 replied:
yessir, compartmentalised into a function!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+
+    def __init__(self, htmlFileFullPath):
+        self.reset()
+        self.results          = set([])
```
If it's a list of actions that needs to be done, name `results` is little confusing to me.

TsubasaK111 replied:
changed to `markers`!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+        prunePos = self.getpos()
+        if self.targetIsFullHTML and tag == 'body':
+            # mark for injection of Google Tag Manager Script.
+            self.results.add(('injectTagManager', prunePos))
```
it's line to inject and not prune, right?
I think `prunePos` should be renamed.

TsubasaK111 replied:
changed to `position`!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+                'en' in self.pruneThisLang and
+                tag == 'a' and
+                attr[0] == 'href' and
+                ('/en/' in attr[1] or attr[1].startswith('en/'))
```
here as well; too long if statement

TsubasaK111 replied:
yessir!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+
+        resultsSet = parser.results.copy()
+        readHtmlFile.close()
+        parser.__kill__()
```
Just question from curiousity, is this 3 lines are for performance issue?

TsubasaK111 replied:
refactored, only `readHtmlFile.close()` remains!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+import os
```
Maybe having shebang would be nice to show what version of python you assume
(I'm not sure the standard of python, so if my suggestion is out of those standards, please disregard)

``` python
#!/usr/bin/env python2
```

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+            for result in resultsSet:
+                resultsList.append(result)
+            resultsList.sort()
+            resultsList.reverse()
```
I'm not sure why this sort and reverse are needed.
Having a comment here would be helpful

TsubasaK111 replied:
This rearranges markers to make modifications from the end of each file.
This prevents the position from becoming outdated.
Also, method chaining (`resultsList.sort().reverse()`) threw an error, hence it is 2 separate instructions.

takayukioda replied:
try `resultsList.sort(reverse=True)`

TsubasaK111 replied:
wow 😍

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+            for result in resultsSet:
+                resultsList.append(result)
+            resultsList.sort()
+            resultsList.reverse()
```
I think this line could be written in one line.
https://docs.python.org/2.7/howto/sorting.html#ascending-and-descending

TsubasaK111 committed:
refactor: rename variables and make logs more verbose

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+                if result[0] == 'injectTagManager':
+                    for line in googleTagManagerScriptLines:
+                        htmlLines.insert(position, line)
+                        position += 1
```
I'm not so sure if you have to split line for `googleTagManagerScript` or not
Can't you just insert one big code in once? Or does htmlLines automatically cut off the lines?

TsubasaK111 replied:
googleTagManagerScript could just be a string, but it just feels wrong...
It's just a gut feeling that I shouldn't have a multi-line string in a List when all the other strings are single line.

TsubasaK111 replied:
`logging.debug` added for tag manager!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+                if result[0] == 'injectTagManager':
+                    for line in googleTagManagerScriptLines:
+                        htmlLines.insert(position, line)
+                        position += 1
```
For consistency, you should have debug log if you put debug log in `prune` action.

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,169 @@
+
+            htmlNewString = ''
+            for line in htmlLines:
+                htmlNewString += (line + '\n')
```
Python have this cool `join` function https://docs.python.org/2/library/string.html#string.join
check it out :sunglasses:

TsubasaK111 replied:
awesome! 😎

takayukioda commented:
@TsubasaK111 maybe you've seen my notes already, but I put some notes so please take a look. fix me

TsubasaK111 commented on .gitignore:
```diff
@@ -1,6 +1,5 @@
-*.py
-*.pyc
+# Logs don't belong.
+*.log
```
おでんが食べたい！

TsubasaK111 committed:
refactor: remove unnecessary code (eg __kill__)

TsubasaK111 committed:
fix: make prune condition more specific

- prevents external links (eg: link to English Node.js documentation) from being pruned

TsubasaK111 committed:
feat: add dateTime information to logs

TsubasaK111 committed:
refactor: compartmentalize long conditionals into functions

TsubasaK111 commented:
@takayukioda mmmm lots of lessons learned, review me!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,168 @@
+                if isExceptionTopIndex(self, attr):
+                    self.markers.remove(('prune', position))
+
+def isPruneLangElement(self, tag, attr):
```
if you place this function out of the class, you shouldn't pass `self`, because function don't know the property of self. instead, give `pruneLangs` as arguement.

TsubasaK111 replied:
will put the functions into the class, where I think they belong!

takayukioda commented on source/build.py:
```diff
@@ -0,0 +1,168 @@
+        parser = PostMkdocsParser(htmlFileFullPath)
+        parser.feed(htmlString)
+
+        readHtmlFile.close()
```
Place this line right after the `htmlString = readHtmlFile.read()` would be nice.
or don't use htmlString and just directly give `readHtmlFile.read()` into `parser.feed` function.

TsubasaK111 replied:
will do!

TsubasaK111 committed:
refactor: relocate helper functions into Parser class

TsubasaK111 committed:
docs: update readme to reflect new build process

takayukioda commented:
![LGTM](https://67.media.tumblr.com/15d756a7e34b48253d8a26bba1f73437/tumblr_o15so0SA9U1rbw3uvo1_500.gif)
